controller/snapshot_controller.go

```diff
@@ -0,0 +1,545 @@
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+	}
+
+	log := getLoggerForSnapshot(sc.logger, snapshot)
+
+	// deleting snapshotCR
+	if !snapshot.DeletionTimestamp.IsZero() {
+		volume, err := sc.ds.GetVolumeRO(snapshot.Spec.Volume)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Infof("Removing finalizer for snapshot %v", snapshot.Name)
+				return sc.ds.RemoveFinalizerForSnapshot(snapshot)
+			}
+			return err
+		}
+		if !volume.DeletionTimestamp.IsZero() {
+			log.Infof("Removing finalizer for snapshot %v", snapshot.Name)
+			return sc.ds.RemoveFinalizerForSnapshot(snapshot)
+		}
+		engine, err := sc.getTheOnlyEngineCRforSnapshot(snapshot)
+		if err != nil {
+			return err
+		}
+		if engine.Status.CurrentState != longhorn.InstanceStateRunning {
+			// TODO: how to prevent duplicated event here
+			sc.eventRecorder.Eventf(snapshot, v1.EventTypeWarning, "SnapshotDeleteError", "cannot delete snapshot because the volume engine %v is not running", engine.Name)
+			sc.logger.Errorf("cannot delete snapshot because the volume engine %v is not running", engine.Name)
+			return nil
+		}
+		// Delete the snapshot from engine process
+		if err := sc.handleSnapshotDeletion(snapshot, engine); err != nil {
+			return err
+		}
+
+		// Wait for the snapshot to be removed from engine.Status.Snapshots
+		engine, err = sc.ds.GetEngineRO(engine.Name)
+		if err != nil {
+			return err
+		}
+		snapshotInfo, ok := engine.Status.Snapshots[snapshot.Name]
+		if !ok {
+			log.Infof("Removing finalizer for snapshot %v", snapshot.Name)
+			return sc.ds.RemoveFinalizerForSnapshot(snapshot)
+		}
+
+		existingSnapshot := snapshot.DeepCopy()
+		if err := syncSnapshotWithSnapshotInfo(snapshot, snapshotInfo, engine.Spec.VolumeSize); err != nil {
+			return err
+		}
+		if reflect.DeepEqual(existingSnapshot.Status, snapshot.Status) {
+			return nil
+		}
+		if _, err = sc.ds.UpdateSnapshotStatus(snapshot); err != nil {
+			return err
+		}
+		sc.generatingEventsForSnapshot(existingSnapshot, snapshot)
+		return nil
+	}
+
+	existingSnapshot := snapshot.DeepCopy()
+	defer func() {
+		if err != nil && !shouldUpdateObject(err) {
+			return
+		}
+		if reflect.DeepEqual(existingSnapshot.Status, snapshot.Status) {
+			return
+		}
+		snapshot, err = sc.ds.UpdateSnapshotStatus(snapshot)
+		if err != nil {
+			return
+		}
+		sc.generatingEventsForSnapshot(existingSnapshot, snapshot)
+		return
+	}()
+
+	engine, err := sc.getTheOnlyEngineCRforSnapshot(snapshot)
+	if err != nil {
+		return err
+	}
+
+	// newly created snapshotCR by user
+	requestCreateNewSnapshot := snapshot.Spec.CreateSnapshot
+	alreadyCreatedBefore := snapshot.Status.CreationTime != nil
+	if requestCreateNewSnapshot && !alreadyCreatedBefore {
+		if engine.Status.CurrentState != longhorn.InstanceStateRunning {
+			snapshot.Status.Error = fmt.Sprintf("cannot take snapshot because the volume engine %v is not running", engine.Name)
+			return nil
+		}
```

Comment:
We only need the engine to be running in 2 cases:
1. snapshot CR  has deletion timestamp and the volume CR is still exist
2. snapshot CR is requesting a new snapshot

I am not sure how can we check the engine one time. Could you give some hints?

Reply:
I mean, I don't need to check `engine.Status.CurrentState != longhorn.InstanceStateRunning` or even get the engine multiple times in one reconcile loop. If there is no sudden change in the engine, the controller just needs to error out then retry the reconciliation

Reply:
Agree with the general logic you mentioned. However, volume not being attached is a common case and I would like to check and generate an event here to let the user know via `kubectl describe` that the snapshot is not being handled because the volume is not attached:
```
			sc.eventRecorder.Eventf(snapshot, v1.EventTypeWarning, "SnapshotDeleteError", "cannot delete snapshot because the volume engine %v is not running", engine.Name)

```

How do you think?